rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,351 @@
+The Google-internal test for TF should catch the error if the rollback sequence
+is not correct.
+
+```python
+# Existing scenario.
+# PR 11111 (2 files updated)
+# +++ tensorflow/python/ops/array_ops.py
+def some_new_function(inputs):
+   ...
+
+# +++ tensorflow/python/keras/layers/core.py
+
+class new_layer(Layer):
+
+  def call(inputs):
+     array_ops.some_new_function(inputs)
+     ...
+```
+
+```python
+# New scenario.
+# PR 22222 (1 file updated)
+# +++ tensorflow/python/ops/array_ops.py
+@tf.export('some_new_function')
+def some_new_function(inputs):
+   ...
+
+==================================
+# PR 33333 (1 file updated)
+# +++ tensorflow/python/keras/layers/core.py
+
+class new_layer(Layer):
+
+  def call(inputs):
+     tf.some_new_function(inputs)
+     ...
+```
+
+2. Changing the behavior of an existing TF API.
+
+Note that the PR 22222 needs to be submitted with both the new and old
+function since Google internal CI is still testing from HEAD.
+The previous function can be 
+deleted after PR 33333 is submitted. Also note that this issue is caused by 
+Keras not using exclusively public TF API, but relying on TF implementation details.
+Moving towards only using public APIs should reduce the likelihood of this kind of issue.
+
+```python
+# Existing scenario.
+# PR 11111 (2 files updated)
+# tensorflow/python/ops/array_ops.py
+<<<
+def existing_function(inputs):
+    ...
+>>>
+def new_function(inputs, knob1=False, knob2=1):
+    ...
+# tensorflow/python/keras/layers/core.py
+
+class existing_layer(Layer):
+
+  def call(inputs):
+<<<
+    array_ops.existing_function(inputs)
+>>>
+    array_ops.new_function(
+        inputs, 
+        knob1=True,
+        knob2=3)
+```
+
+```python
+# New scenario.
+# PR 22222 (1 file updated)
+# tensorflow/python/ops/array_ops.py
+<<<
+def existing_function(inputs):
+   ...
+>>>
+def existing_function(inputs):
+  return new_function(
+    inputs, 
+    knob1=False,
+    knob2=1)
+
+def new_function(inputs, knob1, knob2=1):
+    ...
+
+==================================
+# PR 33333 (1 file updated)
+# tensorflow/python/keras/layers/core.py
+class existing_layer(Layer):
+
+  def call(inputs):
+<<<
+    array_ops.existing_function(inputs)
+     ...
+>>>
+    array_ops.new_function(
+        inputs, 
+        knob1=True,
+        knob2=3)
+```
+
+
+### Performance Implications
+
+There may be some performance implications as we move towards only using
+public TF APIs. We need to maintain a benchmark to ensure that there
+is no performance regression.
+
+### Dependencies
+
+The TensorFlow pip package will auto-install the Keras package, which shouldn't make 
+any difference on the end-user side. Under the hood, Keras will be a different 
+package imported by `tf_core`, like what we do for TF estimator.
+
+### Developer experience Impact
+
+* The local build and test times should be greatly reduced, since compiling TF
+is no longer needed, and Keras is a pure-Python project.
```

Comment:
Can we ditch Bazel and use pytest instead? 😃

Reply:
I think that would be a great change! 
Also return to travis/circleci.

Reply:
Even better...

DOCKER BUILD ALL THE THINGS: https://github.com/tensorflow/addons/pull/1053

Reply:
GitHub Workflow Actions is a good choice as well. From tensorflow SIGs the overall experience is that GitHub Workflow Actions is much better than TravisCI (Workflow Actions might be similar to CircleCI with respect to use experience).

Reply:
@gabrieldemarmiesse, I think at this stage, we can't remove bazel and use pytest since it will affect too much of the Google internal builds (we rely on bazel heavily internally). Wrt to CI choice, I think we are open for that, and I will take a look how github workflow action is setup for other SIGs. 

@seanpmorgan, for tf-addons, how is your current experience with Github workflow actions? Does it meet all the need? eg showing build history, GPU testing, presubmit notification, etc?

Reply:
> I think at this stage, we can't remove bazel and use pytest since it will affect too much of the Google internal builds (we rely on bazel heavily internally).

I understand it's a big task right now, especially since all the test helpers of TF are designed for unittest: https://www.tensorflow.org/api_docs/python/tf/test/TestCase I hope we can work towards being less bazel-dependant in the future, as it's very limiting for python-only depositories. Thanks for your answer!

Reply:
@qlzh727 I can help with GitHub actions if needed. You can see either the actions history either in SIG IO (https://github.com/tensorflow/io/actions) or SIG Addone (https://github.com/tensorflow/addons/actions)

The GPU testing is missing for GitHub Actions as it does not have the GPU machine resource. However, it is still possible to bring self-hosted runner and wire up with GitHub actions. In other words, it could be possible to wire up Kokoro's GPU machine with GitHub action I believe. Some additional effort is needed to make this work though.

Reply:
@qlzh727 It seems we all agree here and this can be discussed after the split since it's an independant subject. Could you minimize this thread? Thanks!